core/src/sumeragi/view_change.rs

```diff
@@ -16,90 +16,50 @@ use crate::block::{EmptyChainHash, VersionedCommittedBlock, VersionedValidBlock}
 /// The proof of a view change. It needs to be signed by f+1 peers for proof to be valid and view change to happen.
 #[derive(Debug, Clone, Decode, Encode, IntoSchema)]
 pub struct Proof {
-    payload: ProofPayload,
-    signatures: SignaturesOf<Self>,
+    pub latest_block_hash: HashOf<VersionedCommittedBlock>,
+    pub view_change_index: u64,
+    pub signatures: Vec<Signature>,
 }
 
 impl Proof {
-    /// Hash of this proof.
-    pub fn hash(&self) -> HashOf<Self> {
-        HashOf::new(&self.payload).transmute()
-    }
-
-    fn from_payload(payload: ProofPayload, key_pair: KeyPair) -> Result<Self> {
-        let signatures = SignaturesOf::new(key_pair, &payload)
-            .wrap_err("Failed to create proof of view change")?
-            .transmute();
-        Ok(Self {
-            payload,
-            signatures,
-        })
-    }
-
-    /// Constructor for `CommitTimeout` view change suggestion.
-    /// # Errors
-    /// Fails if signing failed.
-    pub fn commit_timeout(
-        hash: HashOf<VersionedValidBlock>,
-        previous_proof: HashOf<Self>,
-        latest_block: HashOf<VersionedCommittedBlock>,
-        key_pair: KeyPair,
-    ) -> Result<Self> {
-        let payload = ProofPayload {
-            reason: Reason::CommitTimeout(CommitTimeout { hash }),
-            previous_proof,
-            latest_block,
-        };
-        Self::from_payload(payload, key_pair)
-    }
+    fn signature_payload(&self) -> Hash {
+        use byteorder::{ByteOrder, LittleEndian};
 
-    /// Constructor for `BlockCreationTimeout` view change suggestion.
-    /// # Errors
-    /// Can fail due to signing
-    pub fn block_creation_timeout(
-        previous_proof: HashOf<Self>,
-        latest_block: HashOf<VersionedCommittedBlock>,
-        key_pair: KeyPair,
-    ) -> Result<Self> {
-        let payload = ProofPayload {
-            reason: Reason::from(BlockCreationTimeout),
-            previous_proof,
-            latest_block,
-        };
-        Self::from_payload(payload, key_pair)
-    }
-
-    /// Constructor for `NoTransactionReceiptReceived` view change suggestion.
-    /// # Errors
-    /// Can fail due to signing
-    pub fn no_transaction_receipt_received(
-        previous_proof: HashOf<Self>,
-        latest_block: HashOf<VersionedCommittedBlock>,
-        key_pair: KeyPair,
-    ) -> Result<Self> {
-        let payload = ProofPayload {
-            reason: Reason::NoTransactionReceiptReceived(NoTransactionReceiptReceived),
-            previous_proof,
-            latest_block,
-        };
-        Self::from_payload(payload, key_pair)
+        let mut buf = [0_u8; Hash::LENGTH + std::mem::size_of::<u64>()];
+        let prev_block_hash = self.latest_block_hash.as_ref();
+        for i in 0..Hash::LENGTH {
+            buf[i] = prev_block_hash[i];
+        }
```

Comment:
All of this can be done without `mut`, if you use `iter()`, `cloned()` and `take(Hash::LENGTH + std::mem::size_of::<u64>()`.

Reply:
The added benefit is that you can then use something like `rayon` to do it using SIMD (the Rust compiler isn't smart-enough to do it that way every time).

Reply:
Here's what when through my head when I wrote this code.
1. I need a buffer containing the hash, followed by the view change index in an endian specific order.
2. I will declare a buffer. Copy the hash in to it. Then write the index in little endian at the end.
3. Now write the dumbest code that does exactly that, a loop. Because dumb code is often fast code.

I will change to `copy_from_slice`. Because it will indeed get optimized into a memcpy. But notice what the implentation of `copy_from_slice` looks like.
```Rust
    default fn spec_clone_from(&mut self, src: &[T]) {
        assert!(self.len() == src.len(), "destination and source slices have different lengths");
        // NOTE: We need to explicitly slice them to the same length
        // to make it easier for the optimizer to elide bounds checking.
        // But since it can't be relied on we also have an explicit specialization for T: Copy.
        let len = self.len();
        let src = &src[..len];
        for i in 0..len {
            self[i].clone_from(&src[i]);
        }
    }
```
It's control flow and basic data ops all the way down. Even though the function calls often disguise that fact.